### ISSUE TITLE
HI！about bilinear

### ISSUE THREAD

**Issue by splinter21** [yu45020/Waifu2x#2]
In the readme, you said "Another more effective model is to add upscaled input image to the final convolution. A simple bilinear upscaled image seems sufficient."
But if bilinear will bring the loss of information and make blurred? Also it can't be learnt. I think bilieast+conv would be better.

**Comment by yu45020** [comment]
What's "bilieast" ? [Pytorch supports bilinear](https://pytorch.org/docs/stable/nn.html#torch.nn.Upsample), so I use bilinear. 

PIL offers more up-scaling methods, and it is interesting to compare them, though I suspect they will have eye distinguishable difference. Any up-scaling methods will either loss detailed information or bring in extra noise, so the convolution will hopefully pick up the residual parts. 

I don't have plans to check different up-scaling methods now as I have a very limited GPU. But you are welcome to share your experiment.

**Comment by yu45020** [comment]
@splinter22 
The following plots are test results per epoch. Training images are 240 patches with size 128x128. Test images are 100 patches with the same size. All models have the same initial parameters and configurations.

Bicubic upscale seems have little difference compared to bilinear, thought it might change if I train it longer. 

* feature res: features after the first convolution are added to the layers right before pixel shuffling
* bilinear res: bilinear upscale input image and add to the output of convolution
* bicubic res: bicubic upscale input image 


loss
![loss](https://user-images.githubusercontent.com/28139045/51641524-5cf79280-1f34-11e9-9990-30156edc79a5.png)

ssim
![ssim](https://user-images.githubusercontent.com/28139045/51641525-5d902900-1f34-11e9-82f6-f53309377d4d.png)


psnr 
![psnr](https://user-images.githubusercontent.com/28139045/51641523-5cf79280-1f34-11e9-8720-f7a0a0a108c3.png)

**Comment by splinter21** [comment]
sorry, that's my fault... I originally wanted to say "nearest"...

**Comment by zhouhang95** [comment]
双线性是线性的 理论上只需要一层conv就能恢复 连激活函数都不需要
而二次立方 是非线性的 恢复原有信息就需要经过几层激活函数

**Comment by yu45020** [comment]
@zhouhang95 
有个实际问题：如果模型是用fp16或者混合精度训练的, 数值误差带来的信息损失能否单层就能回复？MobileNet系列取消了部分激活函数来提高准确度，那么恢复2此立方的信息需不需要不同的结构层？

**Comment by zhouhang95** [comment]
单层就能回复 这句话我没理解什么意义 
半精度带来的损失 可以通过加层来恢复 层数越多恢复的越好

MobileNet系列取消了部分激活函数来提高准确度
你要明白为什么取消部分激活函数
因为relu会截断信息从而丢失 想恢复需要加更多的非线性层 
而MobileNet需要更少的算力 所以就去掉了

那么恢复二次立方的信息需不需要不同的结构层
不 就是是加层 
因为你网络能力本来就那么大 
原本只是恢复残差 
现在不仅要恢复残差 而且还要把残差match到经过二次立方放大的图片上

**Comment by zhouhang95** [comment]
https://www.bilibili.com/video/av20961661
感兴趣可以看看这个 理解一下 理论

**Comment by yu45020** [comment]
是我理解错了，“单层就能回复”就不用理了。 

关于MobileNet，我在另一个项目里做image segmentation是，relu换成leaky relu, selu, 精度还是没有去掉activation的来得高，所以我猜想会不会不仅仅是运算力的关系才去掉。 迟点要找时间补下那个系列的论文。

**Comment by zhouhang95** [comment]
你训练用的是几张卡 啥型号的
